diff --git a/third_party/dawn/src/dawn/native/vulkan/BackendVk.cpp b/src/dawn/native/vulkan/BackendVk.cpp
index c34bd5769..f0b71ec32 100644
--- a/third_party/dawn/src/dawn/native/vulkan/BackendVk.cpp
+++ b/third_party/dawn/src/dawn/native/vulkan/BackendVk.cpp
@@ -575,12 +575,6 @@ std::vector<Ref<PhysicalDeviceBase>> Backend::DiscoverPhysicalDevices(
     };
 
     for (ICD icd : kICDs) {
-#if DAWN_PLATFORM_IS(MACOS)
-        // On Mac, we don't expect non-Swiftshader Vulkan to be available.
-        if (icd == ICD::None) {
-            continue;
-        }
-#endif  // DAWN_PLATFORM_IS(MACOS)
         // We always search for fallback adapters first, so if we already found one, don't bother
         // looking for more.
         if (options->forceFallbackAdapter && !physicalDevices.empty()) {
