diff --git a/third_party/dawn/third_party/dawn/src/dawn/native/vulkan/BackendVk.cpp b/third_party/dawn/third_party/dawn/src/dawn/native/vulkan/BackendVk.cpp
index d0d9d3e26..f0b71ec32 100644
--- a/third_party/dawn/third_party/dawn/src/dawn/native/vulkan/BackendVk.cpp
+++ b/third_party/dawn/third_party/dawn/src/dawn/native/vulkan/BackendVk.cpp
@@ -32,6 +32,7 @@
 #include <utility>
 
 #include "dawn/common/Assert.h"
+#include "dawn/common/GPUInfo.h"
 #include "dawn/common/Log.h"
 #include "dawn/common/SystemUtils.h"
 #include "dawn/native/ChainUtils.h"
@@ -164,22 +165,41 @@ constexpr SkippedMessage kSkippedMessages[] = {
      "vkAllocateMemory(): pAllocateInfo->pNext<VkMemoryDedicatedAllocateInfo>"},
     // crbug.com/324282958
     {"NVIDIA", "vkBindImageMemory: memoryTypeIndex"},
-};
+
+    // crbug.com/441788589
+    {"VUID-vkCmdDraw-None-08114",
+     // vkCmdDraw(): the descriptor
+     "is being used in draw but has never been updated via vkUpdateDescriptorSets() or a similar "
+     "call."},
+
+    // This error gets raised when using MSAARenderToSingleSampled with CreateRenderPass2 because
+    // we have a mismatch in the number of samples in the actual render pass vs. the render pass the
+    // graphics pipeline was created with since we lack MSAARenderToSingleSampled info at pipeline
+    // creation time. This mismatch does not have an effect on any known drivers because they don't
+    // rely on the attachment sample count when rendering with MSAARenderToSingleSampled.
+    // Unfortunately this suppression is overly broad because the check in question is bundled with
+    // all the rest of the render pass compatibility rules. Given that this isn't an issue when
+    // using Dynamic Rendering, however, we should be able to remove the suppression if we ever drop
+    // the CreateRenderPass(2) rendering paths. (ie: if we upgrade to requiring Vulkan 1.3)
+    // http://crbug.com/463893793, https://gitlab.khronos.org/vulkan/vulkan/-/issues/4662
+    {"VUID-vkCmdDraw-renderPass-02684", "The current render pass must be compatible"}};
 
 namespace dawn::native::vulkan {
 
 namespace {
 
+// This should always be sorted such that fallback ICDs are searched first to ensure that we return
+// the correct adapters when users are asking for forced fallback adapters.
 static constexpr ICD kICDs[] = {
+#if defined(DAWN_ENABLE_SWIFTSHADER)
+    ICD::SwiftShader,
+#endif  // defined(DAWN_ENABLE_SWIFTSHADER)
 // Other drivers should not be loaded with MSAN because they don't have MSAN instrumentation.
 // MSAN will produce false positives since it cannot detect changes to memory that the driver
 // has made.
 #if !defined(MEMORY_SANITIZER)
     ICD::None,
 #endif
-#if defined(DAWN_ENABLE_SWIFTSHADER)
-    ICD::SwiftShader,
-#endif  // defined(DAWN_ENABLE_SWIFTSHADER)
 };
 
 // Suppress validation errors that are known. Returns false in that case.
@@ -375,10 +395,13 @@ MaybeError VulkanInstance::Initialize(const InstanceBase* instance, ICD icd) {
     DAWN_TRY(mFunctions.LoadGlobalProcs(mVulkanLib));
 
     DAWN_TRY_ASSIGN(mGlobalInfo, GatherGlobalInfo(mFunctions));
-    if (icd != ICD::SwiftShader && mGlobalInfo.apiVersion < VK_MAKE_API_VERSION(0, 1, 1, 0)) {
-        // See crbug.com/850881, crbug.com/863086, crbug.com/1465064, crbug.com/346990068
-        return DAWN_INTERNAL_ERROR(
-            "Vulkan 1.0 driver is unsupported. At least Vulkan 1.1 is required.");
+
+    if (mGlobalInfo.apiVersion < kRequiredVulkanVersion) {
+        std::ostringstream versionError;
+        versionError << "Vulkan " << FormatAPIVersion(mGlobalInfo.apiVersion)
+                     << " driver is unsupported. At least Vulkan "
+                     << FormatAPIVersion(kRequiredVulkanVersion) << " is required.";
+        return DAWN_INTERNAL_ERROR(versionError.str());
     }
 
     VulkanGlobalKnobs usedGlobalKnobs = {};
@@ -437,10 +460,7 @@ ResultOrError<VulkanGlobalKnobs> VulkanInstance::CreateVkInstance(const Instance
     std::vector<const char*> extensionNames;
     for (InstanceExt ext : extensionsToRequest) {
         const InstanceExtInfo& info = GetInstanceExtInfo(ext);
-
-        if (info.versionPromoted > mGlobalInfo.apiVersion) {
-            extensionNames.push_back(info.name);
-        }
+        extensionNames.push_back(info.name);
     }
 
     VkApplicationInfo appInfo;
@@ -450,7 +470,7 @@ ResultOrError<VulkanGlobalKnobs> VulkanInstance::CreateVkInstance(const Instance
     appInfo.applicationVersion = 0;
     appInfo.pEngineName = "Dawn";
     appInfo.engineVersion = 0;
-    appInfo.apiVersion = std::min(mGlobalInfo.apiVersion, VK_API_VERSION_1_3);
+    appInfo.apiVersion = kRequiredVulkanVersion;
 
     VkInstanceCreateInfo createInfo;
     createInfo.sType = VK_STRUCTURE_TYPE_INSTANCE_CREATE_INFO;
@@ -544,8 +564,20 @@ std::vector<Ref<PhysicalDeviceBase>> Backend::DiscoverPhysicalDevices(
     const UnpackedPtr<RequestAdapterOptions>& options) {
     std::vector<Ref<PhysicalDeviceBase>> physicalDevices;
     InstanceBase* instance = GetInstance();
+
+    auto IsFallbackAdapter = [](const PhysicalDevice* physicalDevice) {
+        // Swiftshader is the only fallback adapter that we currently have.
+        if (gpu_info::IsGoogleSwiftshader(physicalDevice->GetVendorId(),
+                                          physicalDevice->GetDeviceId())) {
+            return true;
+        }
+        return false;
+    };
+
     for (ICD icd : kICDs) {
-        if (options->forceFallbackAdapter && icd != ICD::SwiftShader) {
+        // We always search for fallback adapters first, so if we already found one, don't bother
+        // looking for more.
+        if (options->forceFallbackAdapter && !physicalDevices.empty()) {
             continue;
         }
         if (mPhysicalDevices[icd].empty()) {
@@ -573,6 +605,9 @@ std::vector<Ref<PhysicalDeviceBase>> Backend::DiscoverPhysicalDevices(
                 if (instance->ConsumedErrorAndWarnOnce(physicalDevice->Initialize())) {
                     continue;
                 }
+                if (options->forceFallbackAdapter && !IsFallbackAdapter(physicalDevice.Get())) {
+                    continue;
+                }
                 // This loop can't filter adapters based on SupportsFeatureLevel() since the results
                 // are cached for subsequent calls that might have a different feature level.
                 mPhysicalDevices[icd].push_back(std::move(physicalDevice));
