diff --git a/third_party/spirv-cross/spirv_glsl.cpp b/third_party/spirv-cross/spirv_glsl.cpp
index 2ccc1a472..88a82c9aa 100644
--- a/third_party/spirv-cross/spirv_glsl.cpp
+++ b/third_party/spirv-cross/spirv_glsl.cpp
@@ -176,11 +176,23 @@ void CompilerGLSL::init()
 	const struct lconv *conv = localeconv();
 	if (conv && conv->decimal_point)
 		current_locale_radix_character = *conv->decimal_point;
-#elif defined(__ANDROID__) && __ANDROID_API__ < 26
-	// nl_langinfo is not supported on this platform, fall back to the worse alternative.
-	const struct lconv *conv = localeconv();
-	if (conv && conv->decimal_point)
-		current_locale_radix_character = *conv->decimal_point;
+#elif defined(__ANDROID__)
+    #if __ANDROID_API__ >= 26
+    // nl_langinfo is not even defined if __ANDROID_API__ is not at least 26!
+    if (__builtin_available(android 26, *)) {
+        // localeconv, the portable function is not MT safe ...
+        const char *decimal_point = nl_langinfo(RADIXCHAR);
+        if (decimal_point && *decimal_point != '\0')
+            current_locale_radix_character = *decimal_point;
+    } else
+    #endif
+    if (__builtin_available(android 21, *)) {
+        const struct lconv *conv = localeconv();
+        if (conv && conv->decimal_point)
+            current_locale_radix_character = *conv->decimal_point;
+    } else {
+        // we do nothing
+    }
 #else
 	// localeconv, the portable function is not MT safe ...
 	const char *decimal_point = nl_langinfo(RADIXCHAR);
